app/registry.hs

```diff
@@ -405,7 +411,25 @@ validateURI s =
 validatePledgeAddress :: MonadFail m => String -> m PledgeAddress
 validatePledgeAddress s =
   PledgeAddress <$>
-    decodeBech32 "Pledge address" addrHRP (pack s)
+    decodeBech32 "Pledge address" addrHRP isAddress (pack s)
+  where
+    isAddress bech32@(Bech32 _ datapart) = case Bech32.dataPartToBytes datapart of
+      Nothing -> fail "Couldn't decode address internal payload!?"
+      Just bytes -> bech32 <$ case first BS.unpack (BS.splitAt 1 bytes) of
+        (discriminant, rest) | discriminant == [0x83] ->
+          unless (BS.length rest == pubkeyLength) $ fail
+            "Invalid payload for a 'single' address."
+        (discriminant, rest) | discriminant == [0x84] ->
+          unless (BS.length rest == 2*pubkeyLength) $ fail
+            "Invalid payload for a 'grouped' address."
+        (discriminant, rest) | discriminant == [0x85] ->
+          unless (BS.length rest == pubkeyLength) $ fail
+            "Invalid payload for an 'account' address."
+        (discriminant, rest) | discriminant == [0x86] ->
+          unless (BS.length rest == pubkeyLength) $ fail
+            "Invalid payload for a 'multisig' address."
+        _ ->
+          fail "Unrecognized network discriminant for the given address."
```

Comment:
@KtorZ, can we add a comment about where these choices come from?  Seems far from obvious..

Reply:
Right, this comes from the on-chain binary format used by Jörmungandr (https://github.com/input-output-hk/chain-libs/blob/master/chain-impl-mockchain/doc/format.abnf#L157-L161)

I'll add comment.

Reply:
Fixed in: https://github.com/input-output-hk/testnet-stake-pool-registry-validator/compare/e25c8c7763514db873280f9776ec4a4241f27f56..6e8fb2cfee0fbe5439978a8fa271c5b7c18a28e9